Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerMSE.cpp

```diff
@@ -2129,6 +2292,103 @@ static void appendPipelineAppSinkCapsChanged(GObject*, GParamSpec*, AppendPipeli
     g_timeout_add(0, appSinkCapsChangedFromMainThread, ap);
 }
 
+static GstPadProbeReturn appendPipelineAppsrcDataLeaving(GstPad*, GstPadProbeInfo* info, AppendPipeline* appendPipeline)
+{
+    if (GST_PAD_PROBE_INFO_TYPE(info) & GST_PAD_PROBE_TYPE_BUFFER) {
+        GstBuffer* buffer = GST_PAD_PROBE_INFO_BUFFER(info);
+        gsize bufferSize = gst_buffer_get_size(buffer);
+
+        TRACE_MEDIA_MESSAGE("buffer of size %" G_GSIZE_FORMAT " going thru", bufferSize);
+
+        if (bufferSize > 0)
+            appendPipeline->reportEndOfAppendDataMarkNeeded();
+
+        return GST_PAD_PROBE_OK;
+    }
+
+#ifdef DEBUG_APPEND_PIPELINE_PADS
+    if (GST_PAD_PROBE_INFO_TYPE(info) & GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) {
+        GstEvent* event = GST_PAD_PROBE_INFO_EVENT(info);
+        if (GST_EVENT_TYPE(event) != GST_EVENT_CUSTOM_DOWNSTREAM)
+            return GST_PAD_PROBE_OK;
+
+        const GstStructure* structure = gst_event_get_structure(event);
+        if (!gst_structure_has_name(structure, "end-of-append-data-mark"))
+            return GST_PAD_PROBE_OK;
+
+        guint id = gst_event_get_seqnum(event);
+        TRACE_MEDIA_MESSAGE("custom downstream event id=%u", id);
+
+        return GST_PAD_PROBE_OK;
+    }
+#endif
+
+    ASSERT_NOT_REACHED();
+    return GST_PAD_PROBE_OK;
+}
+
+static GstPadProbeReturn appendPipelineAppsinkDataEntering(GstPad*, GstPadProbeInfo* info, AppendPipeline* appendPipeline)
+{
+    if (GST_PAD_PROBE_INFO_TYPE(info) & GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) {
+        GstEvent* event = GST_PAD_PROBE_INFO_EVENT(info);
+        if (GST_EVENT_TYPE(event) != GST_EVENT_CUSTOM_DOWNSTREAM)
+            return GST_PAD_PROBE_OK;
+
+        const GstStructure* structure = gst_event_get_structure(event);
+        if (!gst_structure_has_name(structure, "end-of-append-data-mark"))
+            return GST_PAD_PROBE_OK;
+
+        guint id = gst_event_get_seqnum(event);
+
+        TRACE_MEDIA_MESSAGE("id=%u", id);
+
+        appendPipeline->reportEndOfAppendDataMarkReceived(id);
+
+        return GST_PAD_PROBE_OK;
+    }
+
+#ifdef DEBUG_APPEND_PIPELINE_PADS
+    if (GST_PAD_PROBE_INFO_TYPE(info) & GST_PAD_PROBE_TYPE_BUFFER) {
+        GstBuffer* buffer = GST_PAD_PROBE_INFO_BUFFER(info);
+        TRACE_MEDIA_MESSAGE("buffer of size %" G_GSIZE_FORMAT " going thru", gst_buffer_get_size(buffer));
+        return GST_PAD_PROBE_OK;
+    }
+#endif
+
+    ASSERT_NOT_REACHED();
+    return GST_PAD_PROBE_OK;
+}
+
+#ifdef DEBUG_APPEND_PIPELINE_PADS
+static GstPadProbeReturn appendPipelinePadProbeDebugInformation(GstPad*, GstPadProbeInfo* info, struct PadProbeInformation* padProbeInformation)
+{
+    ASSERT(GST_PAD_PROBE_INFO_TYPE(info) != static_cast<GstPadProbeType>(GST_PAD_PROBE_TYPE_BUFFER | GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM));
```

Comment:
I don't understand the purpose of this assert condition. Is it ever possible that a probe is called for "something" that is a buffer and an event at the same time? (the only case that would trigger the assertion)